saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/logout/Saml2LogoutAuthenticatorResult.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.provider.service.authentication.logout;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+
+import org.springframework.security.saml2.core.Saml2Error;
+import org.springframework.util.Assert;
+
+/**
+ * A result emitted from a SAML 2.0 Response validation attempt
+ *
+ * @author Josh Cummings
+ * @since 5.6
+ */
+public final class Saml2LogoutAuthenticatorResult {
+
+	static final Saml2LogoutAuthenticatorResult NO_ERRORS = new Saml2LogoutAuthenticatorResult(Collections.emptyList());
+
+	private final Collection<Saml2Error> errors;
+
+	private Saml2LogoutAuthenticatorResult(Collection<Saml2Error> errors) {
+		Assert.notNull(errors, "errors cannot be null");
+		this.errors = new ArrayList<>(errors);
+	}
+
+	/**
+	 * Say whether this result indicates success
+	 * @return whether this result has errors
+	 */
+	public boolean hasErrors() {
+		return !this.errors.isEmpty();
+	}
+
+	/**
+	 * Return error details regarding the validation attempt
+	 * @return the collection of results in this result, if any; returns an empty list
+	 * otherwise
+	 */
+	public Collection<Saml2Error> getErrors() {
+		return Collections.unmodifiableCollection(this.errors);
+	}
+
+	/**
+	 * Return a new {@link Saml2LogoutAuthenticatorResult} that contains both the given
+	 * {@link Saml2Error} and the errors from the result
+	 * @param error the {@link Saml2Error} to append
+	 * @return a new {@link Saml2LogoutAuthenticatorResult} for further reporting
+	 */
+	public Saml2LogoutAuthenticatorResult concat(Saml2Error error) {
```

Comment:
I'm curious where the convention of concat comes from?

Reply:
It seems we lack a simple way to add multiple errors without creating additional `Saml2LogoutAuthenticatorResult` instances. I wonder if it would be better to have a `mutate()` method and then the builder would have addError and/or addErrors method. This isolates the Saml2LogoutAuthenticatorResult from the logic for creating a new one based on `this`. It is also a pattern that is more reproducible for other domain objects (in fact Framework does this in places like `ServerWebExchange`) which would give us more consistency.